Sources/Hub/HubApi.swift

```diff
@@ -0,0 +1,140 @@
+        
+        var source: URL {
+            // https://huggingface.co/coreml-projects/Llama-2-7b-chat-coreml/resolve/main/tokenizer.json?download=true
+            var url = URL(string: "https://huggingface.co")!
+            if repoType != .models {
+                url = url.appending(component: repoType.rawValue)
+            }
+            url = url.appending(path: repoId)
+            url = url.appending(path: "resolve/main")  // TODO: revisions
+            url = url.appending(path: relativeFilename)
+            return url
+        }
+        
+        var destination: URL {
+            repoDestination.appending(path: relativeFilename)
+        }
+        
+        var downloaded: Bool {
+            FileManager.default.fileExists(atPath: destination.path)
+        }
+        
+        func prepareDestination() throws {
+            let directoryURL = destination.deletingLastPathComponent()
+            try FileManager.default.createDirectory(at: directoryURL, withIntermediateDirectories: true, attributes: nil)
+        }
+
+        // Note we go from Combine in Downloader to callback-based progress reporting
+        // We'll probably need to support Combine as well to play well with Swift UI
+        // (See for example PipelineLoader in swift-coreml-diffusers)
+        @discardableResult
+        func download(progressHandler: @escaping (Double) -> Void) async throws -> URL {
+            guard !downloaded else { return destination }
+            
+            try prepareDestination()
+            let downloader = Downloader(from: source, to: destination)
+            let downloadSubscriber = downloader.downloadState.sink { state in
+                if case .downloading(let progress) = state {
+                    progressHandler(progress)
+                }
+            }
+            // We need to assign the cancellable to a var so we keep receiving events, so we suppress the "unused var" warning here
+            let _ = downloadSubscriber
+            try downloader.waitUntilDone()
+            return destination
+        }
+    }
+    
+    func snapshot(from repoId: String, repoType: RepoType = .models, matching glob: String? = nil, progressHandler: @escaping (Progress) -> Void) async throws -> URL {
+        let filenames = try await getFilenames(from: repoId, repoType: repoType, matching: glob)
+        let progress = Progress(totalUnitCount: Int64(filenames.count))
+        let repoDestination = destination(repoId: repoId, repoType: repoType)
+        for filename in filenames {
+            let fileProgress = Progress(totalUnitCount: 100, parent: progress, pendingUnitCount: 1)
+            let downloader = HubFileDownloader(repoId: repoId, repoType: repoType, repoDestination: repoDestination, relativeFilename: filename)
+            try await downloader.download { fractionDownloaded in
+                fileProgress.completedUnitCount = Int64(100 * fractionDownloaded)
+                progressHandler(progress)
```

Comment:
Note we are sending the total progress to the callback, where the number of progress units is `100 * filenames.count`, which is not accurate for a mix of small and large files. We keep track of the current file download progress in a child `fileProgress` instance. We could send both to the callback for more accurate reporting. We could also retrieve file metadata (including size) from the API.